### PR TITLE
Update New Game route & new_game in Nim Class. Add route Unit Tests

### PR DESCRIPTION
**Pull request by daveLambe**
Update New Game route & new_game in Nim Class. Add route Unit Tests in test_routes.py. 
Now accepting and validating JSON to create a new game.

### CODE REVIEW

File: nim/app.py [hunk 1]
```diff
@@ -34,12 +36,15 @@ def update_game_state():
...
-    json = loads('{"min": 3, "max": 12, "piles": 3}')
-    state = app.game.new_game(
-        min=json['min'],
-        max=json['max'],
-        piles=json['piles'],
-    )
+    game_request = request.get_json()
+    try:
+        state = app.game.new_game(
+            min=game_request['min'],
```
Comment (dodecaphonix) [inline nim/app.py]: We can get a key error here the the JSON is incomplete.
`min=game_request.get('min') or session.get('min') or DEFAULT_MIN`
Same for `'max'` and `'piles'`

Comment (romainrichard) [reply nim/app.py]: I think part of that logic should be in `new_game()`. I don't want to always pass `min`, `max` or `piles`, so they should take a default value.
Once that's done, here we could do:
```
    try:
        params = {param: game_request[param]
                  for param in ("min", "max", "piles")
                  if param in game_request}
        state = app.game.new_game(**params)
    except NimException as e:
        return jsonify({'error': e.args})
```
I'm not sure if we should support passing params through the `session`. Why do you want to add that?

File: nim/tests/test_routes.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
+from nim.app import app
+import pytest
+import json
+
+
+@pytest.fixture
+def test_app():
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: May be cleaner to call this `client` to avoid repeating `test_*` more than necessary.

File: nim/tests/test_routes.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    return app.test_client()
+
+
+def test_index(test_app):
+    response = test_app.get('/', content_type='html/text')
+    assert response.status_code == 200
+
+
+def test_new_success(test_app):
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: Most of the functions below should be factorized, it should make it much easier to test all those edge cases.

File: nim/tests/test_routes.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    assert json.loads(new_game_response.data) == expected_json
+
+
+def test_new_max_limit_fail(test_app):
+    new_game_response = test_app.post(
+        '/new',
+        data='{"min": 10, "max": 51, "piles": 4}',
+        content_type='application/json'
+    )
+    expected_json = {"error": ["Please use integers between 1 and 50"]}
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: Checking the human-readable error messages isn't a good idea. Instead of just returning `error`, we should probably have a code with it, maybe `error_code`. And you'd check the value of that code here instead, that way we'll be able to change what we display to the user without breaking out tests.

File: nim/tests/test_routes.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    response = test_app.get('/', content_type='html/text')
+    assert response.status_code == 200
+
+
+def test_new_success(test_app):
+    new_game_response = test_app.post(
+        '/new',
+        data='{"min": 10, "max": 10, "piles": 4}',
+        content_type='application/json'
+    )
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: You should also check the `status_code`.

File: nim/tests/test_routes.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    return app.test_client()
+
+
+def test_index(test_app):
+    response = test_app.get('/', content_type='html/text')
+    assert response.status_code == 200
+
+
+def test_new_success(test_app):
+    new_game_response = test_app.post(
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: `response` should be enough, we'll be using that everywhere so might as well start by being consistent.

File: nim/app.py (hunk 1 above)
Comment (dodecaphonix) [reply nim/app.py]: If we refresh the page everything will go back to default values. If we store their preferences in the session, then people can refresh the page, close and reopen, &c without having to reset all of their preferences every time. This is some super necessary behavior for the front end.

At this point we are only passing game preferences so that comprehension isn't doing anything.
```
In [1]: foo = {'a': 1, 'b': 2}

In [2]: bar = {param: foo[param] for param in ('a', 'b', 'c') if param in foo}

In [3]: bar
Out[3]: {'a': 1, 'b': 2}

In [4]: bar == foo
Out[4]: True
```
 I do like your idea of delegating logic to `new_game`. Since when we unpack dictionaries, the second overwrites the first, we can do...
```
params = {**session.get('game_prefs', {}), **game_request}
state = app.game.new_game(**params)
```
And if `game_request` ever becomes more complicated...
```
params = {**session.get('game_prefs', {}), **game_request.get('game_prefs', {})}
state = app.game.new_game(**params)
```

File: nim/app.py [hunk 7]
```diff
@@ -34,12 +36,15 @@ def update_game_state():
...
+            max=game_request['max'],
+            piles=game_request['piles'],
+        )
+    except NimException as e:
+        return jsonify({'error': e.args})
```
Comment (dodecaphonix) [inline nim/app.py]: We have to tell the front end something went wrong. I'd tell them that they made a bad request. The code for that is 400. Here's how you change your code from the default 200 to a 400.
`return jsonify({'error': e.args}), 400`

File: nim/tests/test_routes.py (hunk 4 above)
Comment (dodecaphonix) [reply nim/tests/test_routes.py]: Romain's comment here goes hand in hand with the comment that I made about returning a 400.

File: nim/app.py (hunk 1 above)
Comment (romainrichard) [reply nim/app.py]: Holy fuck that's beautiful! So your brain is functioning at 100% even at 5am huh?

Comment (dodecaphonix) [reply nim/app.py]: Hahahahahahaha :)

Comment (daveLambe) [reply nim/app.py]: So I first looked at @dodecaphonix code yesterday. I knew then it was beautiful.. but I didn't know how much I didn't know.
I was almost entirely unfamiliar with unpacking using * or ** and what you could do with it, especially in the case of dictionaries. After looking up examples, wonderful conversations & help from the man himself (lookin' at you @dodecaphonix ) & playing around with them a bit today, I am starting to grasp just how wonderful this code is. 
I've started implementing this into the new_game() route in Nim and so far it's working beautifully. 
Thanks to both of you for the feedback! 
More to follow 🙂

Comment (daveLambe) [reply nim/app.py]: @dodecaphonix I do have one question regarding `params = {**session.get('game_prefs', {}), **game_request}` 
Why are we setting an empty dict if there is no `game_prefs` ? 
Is there a difference between game_prefs and the defaults we want to take the place of any field not passed? 
Because if they are the same, game_prefs will be a static line set in the code, which would never be empty. We could then update it to whatever when the user sends the arguments. 
Unless we think the user could put this in a bad state, why would we need to be afraid of this ever being blank? 

Alternatively, if game_prefs *is* a different thing to defaults set if there is an unnamed parameters arguments, I see why you're doing this.

File: nim/app.py (hunk 7 above)
Comment (daveLambe) [reply nim/app.py]: Added, as you saw :D

File: nim/tests/test_routes.py (hunk 6 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: Fair enough. My intention was to keep it specific as there will be other routes tested in this file, But since it's in the scope of the function it should be cool.
The fact that 'Response' is an existing thing in flask also made me cautious around that name, but I guess the capital should make it obvious enough that they're different.

Comment (romainrichard) [reply nim/tests/test_routes.py]: Or `resp` if you prefer.
My reasoning is that we'll have to do quite a few of those tests so if we can make it easier to copy/paste or refactor by keeping the names consistent, then that's probably a good thing to do.

File: nim/tests/test_routes.py (hunk 2 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: Changed 🙂

File: nim/tests/test_routes.py (hunk 4 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: I'm a little wary of ONLY checking error code here and stopping there. 
Am I right to be? Is there anything we can do closer to relying on the code for pass/fail but giving some sort of "warning" if the JSON isn't what we expect? 
@dodecaphonix @romainrichard

Comment (romainrichard) [reply nim/tests/test_routes.py]: I'm not sure I'm following.
We need to give an id to identify the status, and a possible error message to give more details.

For example, let's say someone called the endpoint and passed a value for `min` that was greater than `max`.
We could return a `403` (`Forbidden`) error code and an error message saying that `min` needs to be lower than `max`.

Does that help?

File: nim/tests/test_routes.py (hunk 6 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: Yeah I do think you're right. My above arguments are why I didn't do it in the first place, but after thinking about it I think there is no harm in calling it `response` 🙂 I made this change

File: nim/tests/test_routes.py (hunk 3 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: What exactly do you mean by 'factorized'?

File: nim/tests/test_routes.py (hunk 4 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: Ah yes, it does. We're returning an error code along with an error message. Sounds good 🙂 
But in relation to the automation alone, you want the automation to pass or fail entirely dependant on the error code given?

File: nim/tests/test_routes.py (hunk 3 above)
Comment (romainrichard) [reply nim/tests/test_routes.py]: For example, take `test_new_max_limit_fail`, `test_new_piles_limit_fail`, `test_new_all_limit_fail`, `test_new_pile_zero_fail`, `test_new_all_zero_fail`, `test_new_non_int_pile_fail` and `test_new_non_int_all_fail`.
Their code are almost exactly the same, so you could factorize them.

If you're not familiar with the term, is means simplifying your code by creating bits of code that you can re-use. For example, let's say you have the following:
```
def f1(n):
    print(f"Computing {n} + 1")
    res = n + 1
    print(f"{n} + 1 = {res}")
    return res

def f2(n):
    print(f"Returning {n} + 2")
    res = n + 2
    print(f"{n} + 2 = {res}")
    return res

def f3(n):
    print(f"Returning {n} + 3")
    res = n + 3
    print(f"{n} + 3 = {res}")
    return res

def f4(n):
    print(f"Returning {n} + 4")
    res = n + 4
    print(f"{n} + 4 = {res}")
    return res

def f5(n):
    print(f"Returning {n} + 5")
    res = n + 5
    print(f"{n} + 5 = {res}")
    return res
```

That's fine and all, but look at all that code being copied and pasted over and over! There has to be a way to clean that up. Let's factorize our code!

```
def f(n, inc):
    print(f"Returning {n} + {inc}")
    res = n + inc
    print(f"{n} + {inc} = {res}")
    return res

def f1(n):
    return f(n, 1)

def f2(n):
    return f(n, 2)

def f3(n):
    return f(n, 3)

def f4(n):
    return f(n, 4)

def f5(n):
    return f(n, 5)
```

Much cleaner! Now if you need to modify what is being printed by example, you only need to do it in one single place instead of 5 before.

You can apply the same logic to your functions here.

File: nim/tests/test_routes.py (hunk 4 above)
Comment (romainrichard) [reply nim/tests/test_routes.py]: What automation?

Comment (daveLambe) [reply nim/tests/test_routes.py]: You would be hoping to see something like this? 
```
def test_new_max_limit_fail(client_app):
    response = client_app.post(
        '/new',
        data='{"min": 10, "max": 51, "piles": 4}',
        content_type='application/json'
    )
    assert response.status_code == 400
```

@romainrichard

Comment (daveLambe) [reply nim/tests/test_routes.py]: (In terms of how we're checking for an error, not necessarily that this is the exact error code we wanna give in this case)

Comment (romainrichard) [reply nim/tests/test_routes.py]: That's good to check the status, you could go further and check the existence of an error message.
Maybe check that `new_game_response` has the `"error"` key in it and that the length of `new_game_response["error"]` is equal to 1. That will allow us to update the error message without breaking the tests, while still make sure we do send an error message.

File: nim/tests/test_routes.py (hunk 3 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: Cool, that's what I thought you meant but wanted to make sure 🙂 
Nope, I was not familiar with the term

File: nim/tests/test_routes.py (hunk 4 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: Why would the length of `new_game_response["error"]` be 1?

Comment (romainrichard) [reply nim/tests/test_routes.py]: Because in your code, `new_game_response["error"]` is a `list` with one element: the error message. Not sure why it's a `list` but that's another topic.

File: nim/app.py [hunk 8]
```diff
@@ -34,13 +39,19 @@ def update_game_state():
...
-    state = app.game.new_game(
-        min=json['min'],
-        max=json['max'],
-        piles=json['piles'],
-    )
+    """
+    route for /new so user can start new game
+    Uses JSON from frontend for params in new game
+    """
+    session['default_prefs'] = DEFAULT_GAME
```
Comment (romainrichard) [inline nim/app.py]: Whenever someone calls `/new`, you'll overwrite what you had saved in their session by doing this, this rendering the `default_prefs` variable useless. You should remove that line, as well as the `DEFAULT_GAME` variable altogether since you were only using it here.

File: nim/app.py [hunk 9]
```diff
@@ -34,13 +39,19 @@ def update_game_state():
...
+    user_request = request.get_json()
+    game_request = {**session.get('default_prefs', {}), **user_request}
+    try:
+        state = app.game.new_game(**game_request)
+    except NimException as e:
+        return jsonify({'error': e.args}), 400
     session['state'] = state
+    session['default_prefs'] = state
```
Comment (romainrichard) [inline nim/app.py]: This should be `session['default_prefs'] = game_request`, right?

File: nim/game.py [hunk 10]
```diff
@@ -92,6 +95,18 @@ def __init__(self):
         self.chaos = Chaos()
 
     def new_game(self, min, max, piles):
+        """
+        Logic of new game dealt with here.
+        Uses either user-passed JSON or
+        default values to fill in params
```
Comment (romainrichard) [inline nim/game.py]: JSON? Default values?

File: nim/tests/test_routes.py [hunk 11]
```diff
@@ -1,103 +1,110 @@
...
+    :param passed_client: test_client
+    :param req: json to post
+    :return: bool
+    """
+    response = passed_client.post(
+        '/new',
+        content_type='application/json',
+        data=req
+    )
+    assert response.status_code == 200 and not len(json.loads(response.data)['state']) == 0
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: Instead of `not x == 0` you should write `x != 0`.
But also, an empty list is False-y in Python, so you should write:
```
    assert response.status_code == 200 and json.loads(response.data)['state']
```
In this case though, we should check a bit more than that. We have `req` so we know how many piles there should be exactly, and we know the number of stones in each pile should be within a range.
So we should add that too.

File: nim/tests/test_routes.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    :param passed_client: test_client
+    :param req: json to post
+    :return: bool
+    """
+    response = passed_client.post(
+        '/new',
+        content_type='application/json',
+        data=req
+    )
+    assert response.status_code == 400 and b'error' in response.data
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: We also want to make sure we're not returning a state.
And you may be able to do this instead: `and 'error' in json.loads(response.data)` so you don't have to use the `b` prefix.
And it would be nice to check that the error message isn't empty.

File: nim/tests/test_routes.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        content_type='application/json',
+        data=req
+    )
+    assert response.status_code == 200 and not len(json.loads(response.data)['state']) == 0
+
+
+def assert_new_game_failed(passed_client, req):
+    """
+    Asserts post tp /new gets error code 400
+     & the string 'error' exists in the response
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: NIT: Extra space before `&`.

File: nim/tests/test_routes.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+from nim.app import app
+import pytest
+
+
+@pytest.fixture
+def client_app():
+    return app.test_client()
+
+
+def assert_new_game_success(passed_client, req):
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: `passed_`? That sounds strange, of course the argument was passed.
Maybe just `client`?

There may be something better than `req` too. `game_data`?

File: nim/tests/test_routes.py [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    """
+    Asserts post to /new is 200
+    & state is not empty
+    :param passed_client: test_client
+    :param req: json to post
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: We're in Python land, you should be accepting a `dict` instead. Plus now you're forcing all the calls to this function to call `json.dumps()`, clearly a sign that it needs factorizing.

File: nim/tests/test_routes.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+    :return: bool
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: It does not return anything.

File: nim/tests/test_routes.py [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+
+
+def test_new_omit_all_success(client_app):
+    assert_new_game_success(client_app,
+                            json.dumps({}))
+
+
+def test_new_max_limit_fail(client_app):
+    assert_new_game_failed(client_app,
+                           json.dumps({'min': 10, 'max': 51, 'piles': 3}))
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: Since you've shown you can omit some arguments, may be easier to only pass the faulty argument in this function and below.
For example:
```
def test_new_max_limit_fail(client_app):
    assert_new_game_failed(client_app, json.dumps({'max': 51}))
```
Just a thought.

File: nim/tests/test_routes.py [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+
+
+def test_new_piles_limit_fail(client_app):
+    assert_new_game_failed(client_app,
+                           json.dumps({'min': 10, 'max': 10, 'piles': 51}))
+
+
+def test_new_all_limit_fail(client_app):
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: I'm not convinced that `test_new_all_limit_fail` is needed.
What scenario are we protecting ourselves against that isn't already covered by `test_new_max_limit_fail` and `test_new_piles_limit_fail`? That somehow having one of those arguments be 0 would fail but having all of them at 0 would succeed?
I think you can drop this test, or maybe replace it with a test that passes a negative number to make sure we're handling that correctly too.

File: nim/tests/test_routes.py [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+    assert_new_game_failed(client_app,
+                           json.dumps({'min': 51, 'max': 51, 'piles': 51}))
+
+
+def test_new_pile_zero_fail(client_app):
+    assert_new_game_failed(client_app,
+                           json.dumps({'min': 10, 'max': 20, 'piles': 0}))
+
+
+def test_new_all_zero_fail(client_app):
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: This test should also probably be replaced, as far as we know `test_new_pile_zero_fail` failed so this one will fail too because `piles` is 0, thus this test doesn't give us anything.
Instead checking what happens if you just pass 0 for `min` or `max` would give us something.

File: nim/tests/test_routes.py [hunk 20]
```diff
@@ -0,0 +1,110 @@
+import json
+from nim.app import app
+import pytest
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: According to [PEP8](https://www.python.org/dev/peps/pep-0008/#imports), this should read:
```
import json

import pytest

from nim.app import app
```

File: nim/tests/test_routes.py [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+    assert_new_game_failed(client_app,
+                           json.dumps({'min': 30, 'max': 10, 'piles': 4}))
+
+
+def test_new_non_int_pile_fail(client_app):
+    assert_new_game_failed(client_app,
+                           json.dumps({'min': 10, 'max': 20, 'piles': "a"}))
+
+
+def test_new_non_int_all_fail(client_app):
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: Again, this test isn't giving us much information. If you didn't pass `piles` for example it would tell us that we make sure that at least `min` or `max` is an `int`, but with `piles` as it is now we can only say that we check `piles` as the previous test showed.

File: nim/tests/test_routes.py (hunk 5 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: Addressed 🙂

File: nim/app.py (hunk 8 above)
Comment (daveLambe) [reply nim/app.py]: Yeah you're right, I knew that would happen but left it this way because it gave me an easy way to always pass all parameters to new_game. 
Though there are many better ways to do this and this isn't good implementation at all. 
I'll work on this now

File: nim/app.py (hunk 9 above)
Comment (daveLambe) [reply nim/app.py]: Yes, you're totally right. Changing now

File: nim/game.py (hunk 10 above)
Comment (daveLambe) [reply nim/game.py]: I meant that the route is either passing parameters from JSON we get from user action or the default values in `DEFAULT_GAME`

File: nim/tests/test_routes.py (hunk 11 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: Okay 🙂 I'll add that

File: nim/tests/test_routes.py (hunk 17 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: And cover both cases at once? 
My thinking was to deal with one odd case at a time, and not introduce multiple failure points. 
But I see what you're saying.. 
I was concerned that if we do it this way, the ability to only pass one param *OR* the limit itself breaking would cause this test to fail

File: nim/game.py (hunk 10 above)
Comment (romainrichard) [reply nim/game.py]: Except that it isn't. All that `new_game()` knows about is `min`, `max` and `piles`. Here we know nothing about where that data came from. Sure right now our only implementation follows what you said in the comment, but that's only one implementation. That comment doesn't belong here because `new_game()` doesn't know about this at all.

File: nim/tests/test_routes.py (hunk 17 above)
Comment (romainrichard) [reply nim/tests/test_routes.py]: But you already tested above with the `_success` functions that passing a subset of the params work, so here we would now be testing bad values for params.

File: nim/tests/test_routes.py (hunk 18 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: Fair comment. 
The case I had in mind was testing the limit for `min`. Because if you test the limit on `min` without also putting `max` past the limit, you'll get a different error. (`min can't be greater than max').
So I wanted to make sure min reacted as expected to going above limit

File: nim/tests/test_routes.py (hunk 20 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: Done

File: nim/tests/test_routes.py (hunk 18 above)
Comment (romainrichard) [reply nim/tests/test_routes.py]: So at the minimum you should pass a valid number for `piles` then, otherwise all you can say is that passing an invalid value for `piles` fails.

File: nim/app.py (hunk 8 above)
Comment (dodecaphonix) [reply nim/app.py]: We shouldn't be defaulting anything in the route. All game logic should be in game.py.

File: nim/app.py (hunk 9 above)
Comment (dodecaphonix) [reply nim/app.py]: This should be `session['game_prefs']`. It's not a default if it keeps changing and being reset.

File: nim/game.py (hunk 10 above)
Comment (romainrichard) [reply nim/game.py]: That's what I had been trying not to say for days so that Dave would discover it on his own, now it's all spoiled :stuck_out_tongue:

Comment (dodecaphonix) [reply nim/game.py]: Haha, sorry! >_< I'll let you do your thing :)

File: nim/tests/test_routes.py (hunk 18 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: That's fair

File: nim/tests/test_routes.py (hunk 19 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: Changed

File: nim/tests/test_routes.py (hunk 21 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: Good point. Changed

File: nim/app.py [hunk 22]
```diff
@@ -34,12 +39,16 @@ def update_game_state():
...
-        min=json['min'],
-        max=json['max'],
-        piles=json['piles'],
-    )
+    """
+    route for /new so user can start new game
+    Uses JSON from frontend for params in new game
+    """
+    user_request = request.get_json()
+    game_request = {**DEFAULT_GAME, **user_request}
```
Comment (romainrichard) [inline nim/app.py]: That shouldn't be needed. If some values aren't passed when calling `new_game()` , that function should have the right defaults (i.e. you need to modify `new_game()` so it has default values for its arguments).

File: nim/game.py [hunk 23]
```diff
@@ -4,8 +4,16 @@
...
 NUM_LIMIT = 50
-DEFAULT_GAME = {'min': 5, 'max': 15, 'piles': 4}
+DEFAULT_MIN = 5
+DEFAULT_MAX = 20
+DEFAULT_PILES = 3
+DEFAULT_GAME = {
+    'min': DEFAULT_MIN,
+    'max': DEFAULT_MAX,
+    'piles': DEFAULT_PILES,
+    }
```
Comment (romainrichard) [inline nim/game.py]: Pretty sure this last line shouldn't be indented.

File: nim/game.py [hunk 24]
```diff
@@ -92,6 +103,18 @@ def __init__(self):
         self.chaos = Chaos()
 
     def new_game(self, min, max, piles):
+        """
+        Logic of new game dealt with here.
+        Uses min, max & piles to return a
+        list of piles
```
Comment (romainrichard) [inline nim/game.py]: NIT: you don't need to split the line as long as it's not more than 80 characters long (and if you forget, tox should fail). You have other occurrences in your PR where you did this, I won't comment on all of them but you should update them.

File: nim/tests/test_routes.py [hunk 25]
```diff
@@ -0,0 +1,149 @@
...
+
+import pytest
+
+from nim.game import (
+    DEFAULT_MIN,
+    DEFAULT_MAX,
+    DEFAULT_PILES
+)
+
+from nim.app import app
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: The rule isn't to have an empty line between imports.
Here it should look like this:
```
import json

import pytest

from nim.app import app
from nim.game import (
    DEFAULT_MIN,
    DEFAULT_MAX,
    DEFAULT_PILES
)
```
Please refer to [PEP8](https://www.python.org/dev/peps/pep-0008/#imports) for more information.

File: nim/tests/test_routes.py (hunk 14 above)
Comment (romainrichard) [reply nim/tests/test_routes.py]: You seem to have missed this one.

File: nim/tests/test_routes.py [hunk 26]
```diff
@@ -0,0 +1,149 @@
...
+
+def assert_new_game_success(passed_client, req):
+    """
+    Asserts post to /new is 200
+    & state is not empty
+    Checks piles is in line with req &
+    none in response lower than mix or
+    higher than max
+    :param passed_client: test_client
+    :param req: Desired game conditions
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: You missed [this comment](https://github.com/hacksaurz/nim/pull/15#discussion_r174687264).

File: nim/tests/test_routes.py [hunk 27]
```diff
@@ -0,0 +1,149 @@
...
+    )
+    assert response.status_code == 200
+    response = json.loads(response.data)
+    request = json.loads(req)
+    assert 'state' in response
+    state = response['state']
+    assert all(isinstance(x, int) for x in state)
+    assert request.get('piles', DEFAULT_PILES) == len(state)
+    assert request.get('min', DEFAULT_MIN) <= min(state) <= max(state)
+    request.get('max', DEFAULT_MAX)
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: Look at how lonely this line is, not doing anything :stuck_out_tongue:

File: nim/tests/test_routes.py [hunk 28]
```diff
@@ -0,0 +1,149 @@
...
+    none in response lower than mix or
+    higher than max
+    :param passed_client: test_client
+    :param req: Desired game conditions
+    """
+    response = passed_client.post(
+        '/new',
+        content_type='application/json',
+        data=req
+    )
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: Below, you may want to add some inline comments below, maybe some empty lines too so it doesn't look scary to read.

File: nim/tests/test_routes.py [hunk 29]
```diff
@@ -0,0 +1,149 @@
...
+    :param req: json to post
+    """
+    response = passed_client.post(
+        '/new',
+        content_type='application/json',
+        data=req
+    )
+    assert response.status_code == 400
+    assert 'error' in json.loads(response.data) and json.loads(response.data)['error']
+    assert 'state' not in json.loads(response.data)
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: You're potentially calling `json.loads(response.data)` 3 times, let's fix that!
Plus instead of doing `key in dict and dict[key]`, you can use the `.get()` method instead.

File: nim/tests/test_routes.py (hunk 17 above)
Comment (romainrichard) [reply nim/tests/test_routes.py]: Seems like you've missed this comment.
When testing for failure cases, you should only pass what is needed for the test to fail.
For example, if passing `piles=0` should fail, you should NOT also pass `min` and `max`, otherwise we won't know for sure what caused the failure.

File: nim/game.py [hunk 30]
```diff
@@ -5,6 +5,17 @@
 from nim.exceptions import NimException
 
 
+NUM_LIMIT = 50
+DEFAULT_MIN = 5
+DEFAULT_MAX = 20
+DEFAULT_PILES = 3
+DEFAULT_GAME = {
```
Comment (romainrichard) [inline nim/game.py]: Looks like you won't need this variable anymore.

File: nim/tests/test_routes.py (hunk 27 above)
Comment (daveLambe) [reply nim/tests/test_routes.py]: I sawwwww :(

File: nim/app.py [hunk 31]
```diff
@@ -34,12 +33,16 @@ def update_game_state():
...
-        piles=json['piles'],
-    )
+    """
+    route for /new so user can start new game. Uses JSON from frontend
+    for params in new game
+    """
+    user_request = request.get_json()
+    game_request = {**user_request}
+    try:
+        state = app.game.new_game(**game_request)
```
Comment (romainrichard) [inline nim/app.py]: At least for now, I think we could we could simplify this:
```
        state = app.game.new_game(**request.get_json())
```

File: nim/tests/test_routes.py [hunk 32]
```diff
@@ -0,0 +1,131 @@
...
+    assert response.status_code == 200
+    response = json.loads(response.data)
+    assert 'state' in response
+    state = response['state']
+    assert all(isinstance(x, int) for x in state)
+    # Checking num piles returned is requested or default
+    assert request.get('piles', DEFAULT_PILES) == len(state)
+    # Checking piles returned fall within request or default
+    assert ((request.get('min', DEFAULT_MIN) <= min(state) <= max(state)
+            <= request.get('max', DEFAULT_MAX)) <= NUM_LIMIT)
```
Comment (romainrichard) [inline nim/tests/test_routes.py]: You have extra parentheses in there.
Also, Python will convert this to:
```
    assert (request.get('min', DEFAULT_MIN) <= min(state) and
            min(state) <= max(state) and
            max(state) <= request.get('max', DEFAULT_MAX) and
            request.get('max', DEFAULT_MAX) <= NUM_LIMIT)
```
The 2nd test is useless (of course `min(state) <= max(state)`).
May be cleaner to split the tests and also define a new global variable (we really need to find better variable names, but don't worry about it for now):
```
    # Check that the min value was respected
    assert NUM_LIMIT_MIN <= request.get('min', DEFAULT_MIN) <= min(state)
    # Check that the max value was respected
    assert max(state) <= request.get('max', DEFAULT_MAX) <= NUM_LIMIT_MAX
```
We should also have another test function that does something like:
```
def test_constants():
    assert all(NUM_LIMIT_MIN <= x <= NUM_LIMIT_MAX
               for x in (DEFAULT_MIN, DEFAULT_MAX, DEFAULT_PILES))
```

File: nim/app.py (hunk 31 above)
Comment (daveLambe) [reply nim/app.py]: Oh shit, dat nice.

File: nim/app.py [hunk 33]
```diff
@@ -3,11 +3,10 @@
     render_template,
     session,
     Flask,
+    request,
 )
-from json import loads
-
-from game import Nim
-
+from nim.exceptions import NimException
```
Comment (romainrichard) [inline nim/app.py]: NIT: There should be an empty line before this import.

File: nim/game.py [hunk 34]
```diff
@@ -91,7 +97,18 @@ def __init__(self):
         self.greedy = Greedy()
         self.chaos = Chaos()
 
-    def new_game(self, min, max, piles):
```
[the remark this answers is not in the document]
Comment (romainrichard) [reply nim/game.py]: That's a good point.
Let's change:
```
    def new_game(self, min=DEFAULT_MIN, max=DEFAULT_MAX, piles=DEFAULT_PILES):
```
to
```
    @staticmethod
    def new_game(min=DEFAULT_MIN, max=DEFAULT_MAX, piles=DEFAULT_PILES):
```

[the remark this answers is not in the document]
Comment (romainrichard) [reply nim/game.py]: That'll be addressed in a separate ticket.

File: nim/tests/test_routes.py [hunk 35]
```diff
@@ -0,0 +1,142 @@
...
+    is in line with request. Uses arguments passed or DEFAULT
+    :param client: test_client
+    :param request: Desired game conditions
+    """
+    response = client.post(
+        '/new',
+        content_type='application/json',
+        data=json.dumps(request)
+    )
+    assert response.status_code == 200
```
[the remark this answers is not in the document]
Comment (romainrichard) [reply nim/tests/test_routes.py]: I'll see if I can configure codacy to not report that error for our tests, you can ignore that error everywhere it is reported.

File: nim/game.py (hunk 34 above)
[the remark this answers is not in the document]
Comment (daveLambe) [reply nim/game.py]: Can you give me a brief explanation on why we would do it this way? 
So you're using a decorator to make it a static function.. 
I'm guessing by 'method could be function' is that it could be a static global function. 
Does using the decorator instead of moving it out of here make a difference?

[the remark this answers is not in the document]
Comment (romainrichard) [reply nim/game.py]: > Does using the decorator instead of moving it out of here make a difference?

Technically: no.
But since we're creating a new game, it sort of  makes sense to have it in the `Game` class.
Since we're not using `self` in that method, I added the `@staticmethod` to let everyone know that yes it could be a function, but we're keeping it as a method for now and we're ok with it.

I wouldn't worry too much about this because this code will most likely need to change soon anyway. I mostly added this comment for 2 reasons:
1. To make @codacy-bot happy (you're welcome)
2. To introduce you to `@staticmethod`

[the remark this answers is not in the document]
Comment (daveLambe) [reply nim/game.py]: Soooo @staticmethod doesn't actually do anything other than make you write it so that it *could* be a static function? (As in why you have to remove self)

[the remark this answers is not in the document]
Comment (romainrichard) [reply nim/game.py]: I don't fully understand your question.
Maybe this will help? https://docs.python.org/3.6/library/functions.html#staticmethod
